neo/Consensus/ConsensusService.cs

```diff
@@ -225,9 +286,281 @@ private void OnPersistCompleted(Block block)
+                PreparationHash = preparationHash
+            };
+            bool canRestoreView = false;
+            for (int i = 0; i < context.Validators.Length; i++)
+            {
+                if (i == tempContext.PrimaryIndex) continue;
+                if (message.PrepareWitnessInvocationScripts[i] == null) continue;
+                var regeneratedPrepareResponse = tempContext.RegenerateSignedPayload(prepareResponseMsg, (ushort) i,
+                    message.PrepareWitnessInvocationScripts[i], message.PrepareTimestamps[i]);
+                if (!regeneratedPrepareResponse.Verify(snap) || !PerformBasicConsensusPayloadPreChecks(regeneratedPrepareResponse)) continue;
+                prepareResponses.Add((regeneratedPrepareResponse, prepareResponseMsg));
+                // Verify that there are M valid preparations, 1 Prepare Request + (M-1) Prepare responses
+                if (prepareRequest == null || prepareResponses.Count < context.M - 1) continue;
+                canRestoreView = true;
+                break;
+            }
+
+            // Only accept recovery from lower views if there were at least M valid prepare requests
+            if (message.ViewNumber < context.ViewNumber && !canRestoreView) return;
+
+            var verifiedChangeViewWitnessInvocationScripts = new byte[context.Validators.Length][];
+            if (!canRestoreView && message.ChangeViewWitnessInvocationScripts != null)
+            {
+                if (message.ChangeViewWitnessInvocationScripts.Length < context.Validators.Length) return;
+                if (message.ChangeViewTimestamps.Length < context.Validators.Length) return;
+                if (message.OriginalChangeViewNumbers.Length < context.Validators.Length) return;
+
+                var changeViewMsg = new ChangeView { NewViewNumber = message.ViewNumber };
+                int validChangeViewCount = 0;
+                for (int i = 0; i < context.Validators.Length; i++)
+                {
+                    if (message.ChangeViewWitnessInvocationScripts[i] == null) continue;
+
+                    changeViewMsg.ViewNumber = message.OriginalChangeViewNumbers[i];
+                    // Regenerate the ChangeView message
+                    var regeneratedChangeView = tempContext.RegenerateSignedPayload(changeViewMsg, (ushort) i,
+                        message.ChangeViewWitnessInvocationScripts[i], message.ChangeViewTimestamps[i]);
+                    if (!regeneratedChangeView.Verify(snap) || !PerformBasicConsensusPayloadPreChecks(regeneratedChangeView)) continue;
+                    verifiedChangeViewWitnessInvocationScripts[i] = message.ChangeViewWitnessInvocationScripts[i];
+                    validChangeViewCount++;
+                    if (validChangeViewCount < context.M) continue;
+                    canRestoreView = true;
+                    break;
+                }
+            }
+
+            if (!canRestoreView) return;
```

Comment:
Jeff, I felt that this check might be `if (canRestoreView) return;`, but I am not sure.
Because in the last loop you set `canRestoreView = true` if we have `M` valid views.

Reply:
canRestoreView means if we can restore view. So if we can't restore view we return. The logic here is correct. I'm going to resolve this conversation, but feel free to re-open if you believe there is still something wrong with it.